src/main/java/com/ss/mqtt/broker/service/impl/AbstractCredentialSource.java

```diff
@@ -1,22 +1,28 @@
 package com.ss.mqtt.broker.service.impl;
 
 import com.ss.mqtt.broker.service.CredentialSource;
+import com.ss.rlib.common.util.dictionary.ConcurrentObjectDictionary;
+import com.ss.rlib.common.util.dictionary.Dictionary;
+import com.ss.rlib.common.util.dictionary.DictionaryFactory;
+import com.ss.rlib.common.util.dictionary.ObjectDictionary;
 import org.jetbrains.annotations.NotNull;
 import reactor.core.publisher.Mono;
 
-import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
-import java.util.HashMap;
-import java.util.Map;
 
 public abstract class AbstractCredentialSource implements CredentialSource {
 
-    private final Map<String, byte[]> credentials = new HashMap<>();
+    private final ConcurrentObjectDictionary<String, byte[]> credentials =
+        DictionaryFactory.newConcurrentStampedLockObjectDictionary();
```

Comment:
you can use ConcurrentObjectDictionary.of(keyType, valueType) because it provides default concurrent implementation